src/Agda2Hs/Compile/Name.hs

```diff
@@ -147,15 +155,37 @@ compileQName f
     primModules = ["Agda.Builtin", "Haskell.Prim", "Haskell.Prelude"]
     primMonadModules = ["Haskell.Prim.Monad.Dont", "Haskell.Prim.Monad.Do"]
 
-    mkImport mod qual par hf
+    -- Determine whether it is a type operator or an "ordinary" operator.
+    -- _getSort is not for that; e. g. a data has the same sort as its constructor.
+    getNamespace :: QName -> C (Hs.Namespace ())
+    getNamespace qName = do
+      definition <- getConstInfo qName
+      let isQNameSet = isSet $ getResultType $ defType definition
+      reportSDoc "agda2hs.name" 25 $ text $ "Checking whether " ++ (prettyShow $ nameCanonical $ qnameName f)
+                                              ++ "is a type operator: " ++ show isQNameSet
+      if isQNameSet then return (Hs.TypeNamespace ()) else return (Hs.NoNamespace ())
+
+    -- Gets the type of the result of the function (the type after the last "->").
+    getResultType :: Type -> Type
+    getResultType typ = case (unEl typ) of
+      (Pi _ absType) -> getResultType $ unAbs absType
+      _              -> typ
+
+    -- I'm not sure whether this is the correct way to determine whether it's a Set/Prop;
```

Comment:
this does not seem right, you should instead look at the *argument* of the `Sort` constructor to check whether it is a `Set` sort or `Prop` sort. Or even better, use the `isPropM` utility function.

Reply:
Maybe I don't understand something about the concepts themselves. I've tried `isQNameSet <- isPropM $ getResultType $ defType definition` at line 164, but then I got the same strange results as when I looked at levels: in `TypeOperatorImport.hs`, I've got

```hs
import TypeOperatorExport ((&&&), (***), type (***)((:*:)), (<))
```

The sorts themselves were strange. `(:*:)` seemed to have the same sort (`InfS ...`) as `(***)`. And there were type operators with level `Max 0 ...` and normal operators `Max 1 ...`, and then vice versa. Is there something from the theory which I don't understand? Thank you in advance:)

Reply:
> The sorts themselves were strange. (:*:) seemed to have the same sort (InfS ...) as (***). And there were type operators with level Max 0 ... and normal operators Max 1 ..., and then vice versa. Is there something from the theory which I don't understand? Thank you in advance:)

The reason why these sorts are `InfS` is because both symbols are universe-polymorphic:
```agda
data _***_ {i j : Level} (a : Set i) (b : Set j) : Set (i ⊔ j) where
  _:*:_ : a -> b -> a *** b
```
so they both end up in the sort of universe-polymorphic things, which is `InfS ...`. This suggests that just looking at the sort is not really the right thing to do here. I think that you could instead look at the return type of the symbol and check whether it is a sort (`Sort` term constructor) or a datatype (`Def` term constructor). Could you give that a try and let me know how it goes?

Reply:
Thank you:) If I understand well, I think this was my original concept: `getResultType` gets the return type through a recursion and `isSet` decides based on the constructor whether it's a sort. What I'm unsure about is how to handle the other constructors (`Lam`, `Lit` etc.); probably there is no need for type namespace for them.

Reply:
Ah yes, I see now what you were trying to do. In that case I think your implementation is actually fine, I was just confused by the name `isSet` which should more properly be called `isSort`. And actually this helper function already exists in `Agda.Syntax.Internal` (though it returns a `Maybe Sort` instead of a `Bool`).

Reply:
I didn't know about that; thanks! I'm using that now.